ci: shell quoting hygiene + add shellcheck to lint

While reading through the workflow files I noticed a few shell hygiene issues and wanted to propose fixes + a linter to catch future ones.

**Real bug: mis-matched quotes in test.yml**
```
# current — broken quoting (works by accident)
run: echo "RUSTFLAGS="-C debuginfo=0"" >> $GITHUB_ENV
```

The first `"` closes after `RUSTFLAGS=`, leaving `-C` and` debuginfo=0 `as separate unquoted tokens. `echo` joins them with a space, which happens to produce the right output — but it's fragile and confusing.
```
# proposed
run: echo 'RUSTFLAGS=-C debuginfo=0' >> "$GITHUB_ENV"
```

Hygiene: unquoted `$(cmd) `substitutions and intermediate variables
A few run: steps assign `VAR=$(cmd)` and reference `$VAR` on the next line, or use unquoted `$(cmd)` results. While GitHub Actions writes run: blocks to a temp file rather than invoking them as `sh -c "..."`, the patterns are still fragile — a path with spaces or glob characters would silently misbehave.

Specific instances:

`release.yml: ``VERSION=$(echo $GITHUB_REF | sed ...) `then `$VERSION` used (×2, PyPI + crates.io steps)
`release.yml`: `export SDKROOT=$(xcrun ...)` — unquoted substitution
`test.yml`: `python_version=$(echo ... | sed ...)` then `$python_version` used
`test.yml`: `export PYO3_CONFIG_FILE=$(pwd)/... `then used on next line
Proposed fix: inline substitutions directly and quote `$(cmd)` results.

**Enhancement: add shellcheck to the lint job**
`shellcheck` is pre-installed on GitHub-hosted ubuntu runners. Adding it to the existing `lint` job would catch these classes of issue automatically going forward:

```
- name: shellcheck
  run: find . -name "*.sh" | xargs shellcheck --severity=warning
```
One prerequisite: [.devcontainer/post_create.sh](https://github.com/reggswilgoxson/maturin/blob/claude/shell-variable-expansion-audit-Uh6wl/.devcontainer/post_create.sh) is missing a shebang, which causes shellcheck to exit with SC2148 (unknown target shell). Adding `#!/bin/bash` fixes that.

**Happy to open a PR with all of the above if this looks reasonable to you.**

--- a/.devcontainer/post_create.sh
+++ b/.devcontainer/post_create.sh
@@ -1,3 +1,4 @@
+#!/usr/bin/env bash
 set -euxo pipefail
 
 sudo apt-get update

--- a/.github/workflows/release.yml
+++ b/.github/workflows/release.yml
@@ -103,7 +103,8 @@ jobs:
         run: |
           set -euo pipefail
           # set SDKROOT for C dependencies like ring and bzip2
-          export SDKROOT=$(xcrun --sdk macosx --show-sdk-path)
+          SDKROOT="$(xcrun --sdk macosx --show-sdk-path)"
+          export SDKROOT
           rustup target add aarch64-apple-darwin
           cargo run -- build --release -b bin -o dist --target universal2-apple-darwin --features password-storage,static --compatibility pypi
 

--- a/.github/workflows/test.yml
+++ b/.github/workflows/test.yml
@@ -129,11 +129,12 @@ jobs:
 
       - name: Set MATURIN_TEST_PYTHON
         shell: bash
-        run: echo "MATURIN_TEST_PYTHON=${{ steps.setup-python.outputs.python-path }}" >> $GITHUB_ENV
+        run: echo "MATURIN_TEST_PYTHON=${{ steps.setup-python.outputs.python-path }}" >> "${GITHUB_ENV}"
       # To save cache disk space
       - name: Disable debuginfo on Windows
         if: startsWith(matrix.os, 'windows')
-        run: echo "RUSTFLAGS="-C debuginfo=0"" >> $GITHUB_ENV
+        shell: bash
+        run: echo "RUSTFLAGS=-C debuginfo=0" >> "${GITHUB_ENV}"
       - name: cargo test
         run: cargo nextest run --features password-storage
       - name: test cross compiling with zig
@@ -173,7 +174,8 @@ jobs:
         run: |
           set -ex
           rustup target add x86_64-unknown-linux-gnu
-          export PYO3_CONFIG_FILE=$(pwd)/test-crates/pyo3-mixed/pyo3-config.txt
+          PYO3_CONFIG_FILE="$(pwd)/test-crates/pyo3-mixed/pyo3-config.txt"
+          export PYO3_CONFIG_FILE
           target/debug/maturin build -m test-crates/pyo3-mixed/Cargo.toml --target x86_64-unknown-linux-gnu --zig
       - name: test maturin new
         shell: bash

